pkg/services/vimmachine.go

```diff
@@ -393,12 +411,63 @@ func (v *VimMachineService) createOrPatchVSphereVM(ctx context.Context, vimMachi
 
 // generateVMObjectName returns a new VM object name in specific cases, otherwise return the same
 // passed in the parameter.
-func generateVMObjectName(vimMachineCtx *capvcontext.VIMMachineContext, machineName string) string {
+func generateVMObjectName(vimMachineCtx *capvcontext.VIMMachineContext, machineName string) (string, error) {
+	name, err := GenerateVSphereVMName(machineName, vimMachineCtx.VSphereMachine.Spec.NamingStrategy)
+	if err != nil {
+		return "", err
+	}
 	// Windows VM names must have 15 characters length at max.
 	if vimMachineCtx.VSphereMachine.Spec.OS == infrav1.Windows && len(machineName) > 15 {
-		return strings.TrimSuffix(machineName[0:9], "-") + "-" + machineName[len(machineName)-5:]
+		return strings.TrimSuffix(name[0:9], "-") + "-" + name[len(name)-5:], nil
 	}
-	return machineName
+	return name, nil
+}
+
+const (
+	maxNameLength = 63
+)
+
+var nameTemplateFuncs = map[string]any{
+	"trimSuffix": sprig.GenericFuncMap()["trimSuffix"],
+	"trunc":      sprig.GenericFuncMap()["trunc"],
+}
+
+var nameTpl = template.New("name generator").Funcs(nameTemplateFuncs).Option("missingkey=error")
+
+// GenerateVSphereVMName generates the name of a VSphereVM based on the naming strategy.
+func GenerateVSphereVMName(machineName string, namingStrategy *infrav1.VSphereVMNamingStrategy) (string, error) {
```

Comment:
This function seems to be the same as we have for supervisor?

Let's move this to `pkg/util` to a separate go file and use it at both places.

Reply:
I kept different because we have different struct for each mode. 

If we want to combine, should we also use a common struct for both then it will be make the pkg/util function more simpler otherwise we might need to do type conversion or use an interface

Reply:
Let's not use the same struct, as these are API types and we need to be very careful about them.

Instead use a common function which takes a `*string` as template (so we only have to check that the template is not nil).